lib/FetchEnvs.js

```diff
@@ -14,210 +14,291 @@
-        case 'json':
-          result = JSON.parse(result);
-          break;
-        case 'jsonString':
-          // Stringify the jsonstring. This has the effect of double escaping the json, so that
-          // when we go to parse the final template to apply it to kube, it doesnt mistakenly
-          // turn our jsonString into actual json.
-          result = JSON.stringify(result);
-          // JSON.stringify adds quotes around the newly created json string. Kube forces us
-          // to wrap out curly braces in quotes so that it wont error on our templates. In order
-          // to avoid having 2 double quotes around the result, we need to remove the stringify
-          // quotes. slice(start of slice, end of slice)
-          result = result.slice(1, result.length - 1);
-          break;
-        case 'base64':
-          result = Buffer.from(result).toString('base64');
-          break;
+
+    const data = resource?.data;
+    if (kubeError === undefined) kubeError = 'no data returned from processEnvFrom. make sure your data exists in the correct location and is in the expected format.';
+
+    if (!data) {
+      if (!optional) {
+        throw new Error(`envFrom.genericMapRef.${apiVersion}.${kind}.${name}.data not found: ${kubeError}`);
+      } 
+
+      let msg = `envFrom.genericMapRef.${apiVersion}.${kind}.${name}.data not found: ${kubeError}`;
+      log.warn(msg);
+      this.updateRazeeLogs('warn', { controller: 'FetchEnvs', message: msg });
+    }
+
+    return {
+      ...conf,
+      data
+    };
+  }
+
+  async #genericKeyRef(conf) {
+    const ref = conf.valueFrom.genericKeyRef;
+    const strategy = conf.overrideStrategy;
+    const {
+      name,
+      namespace = this.namespace,
+      kind,
+      key,
+      matchLabels,
+      apiVersion,
+      type
+    } = ref;
+
+    const krm = await this.kubeClass.getKubeResourceMeta(
+      apiVersion,
+      kind,
+      'update'
+    );
+
+    const qs = labelSelectors(matchLabels);
+    const uri = krm.uri({namespace, kind, name});
+    const response = await this.api({
+      uri: uri,
+      json: true,
+      qs: qs
+    });
+
+    if (matchLabels) {
+      const output = response.items.reduce(reduceItemList(ref, strategy), {});
+      return output[key];
+    }
+
+    return typeCast(response?.data?.[key], type);
+  }
+
+  async #secretMapRef(conf) {
+    const name = conf.secretMapRef?.name;
+    const namespace = conf.secretMapRef?.namespace ?? this.namespace;
+    const optional = !!conf.optional;
+    let kubeError;
+
+    let res;
+
+    try {
+      res = await this.api({
+        uri: `/api/v1/namespaces/${namespace}/secrets/${name}`,
+        json: true
+      });
+    } catch (error) {
+      log.warn(error);
+      kubeError = error.message;
+    }
+
+    const data = res?.data;
+    if (kubeError === undefined) kubeError = 'no data returned from processEnvFrom. make sure your data exists in the correct location and is in the expected format.';
```

Comment:
This error message is the same everywhere, it could be pulled out and just in the entry point

Reply:
`#processEnvFrom` could be simple for loop rather than a Promise.all, then you can have the single try / catch

Reply:
if you move the try catch out into a for loop, would you also move the `optional` code?

Reply:
Ya, thats fair. It may be better to leave it like this. It currently runs the operations in parallel. Making it a loop makes it synchronous.

I think it may just be the wording on the kubeError message. I'm not sure its all that helpful. As an end user reading these logs wouldn't know what `processEnvFrom` was  the `envFrom.configMapRef.name` is the thing that is important here. 

From this context I don't think the error is really needed.